tests/unit/test_charm.py

```diff
@@ -1,75 +1,300 @@
+                "juju-password": "",
+                "scrape-interval": 5,
+                "scrape-port": 5000,
+            }
+        )
+
+    snap_config = harness.charm.generate_exporter_config()
+
+    for section, missing_keys in expected_missing_config.items():
+        for key in missing_keys:
+            assert not snap_config[section][key]
+
+    for section, present_keys in expected_present_config.items():
+        for key in present_keys:
+            assert snap_config[section][key]
+
+
+def test_reconfigure_scrape_target_success(harness, mocker):
+    """Test updating scrape target of Prometheus successfully."""
+    port = 5000
+    interval_min = 5
+    interval_sec = interval_min * 60
+    timeout = 30
+    expose_target_mock = mocker.patch.object(
+        harness.charm.prometheus_target, "expose_scrape_target"
+    )
+
+    with harness.hooks_disabled():
+        harness.update_config(
+            {"scrape-port": port, "scrape-interval": interval_min, "scrape-timeout": timeout}
+        )
+
+    harness.charm.reconfigure_scrape_target()
+    expose_target_mock.assert_called_once_with(
+        port, "/metrics", scrape_interval=f"{interval_sec}s", scrape_timeout=f"{timeout}s"
+    )
+
+
+def test_reconfigure_scrape_target_fail(harness, mocker):
+    """Test failure when updating scrape target of Prometheus."""
+    expose_target_mock = mocker.patch.object(
+        harness.charm.prometheus_target, "expose_scrape_target"
+    )
+    logger_mock = mocker.patch.object(charm, "logger")
+
+    # re-raise error in case the prometheus target configuration fails
+    exception = charm.PrometheusConfigError()
+    expose_target_mock.side_effect = exception
+    with pytest.raises(charm.PrometheusConfigError):
+        harness.charm.reconfigure_scrape_target()
+
+    logger_mock.error.assert_called_once_with(
+        "Failed to configure prometheus scrape target: %s",
+        exception
+    )
+
+
+def test_reconfigure_open_ports(harness, mocker):
+    """Test updating which ports are open on units."""
+    old_port_spec = "5000/tcp"
+    old_port, old_protocol = old_port_spec.split("/")
+    new_port = 6000
+
+    mocker.patch.object(charm.hookenv, "opened_ports", return_value=[old_port_spec])
+    mock_open_port = mocker.patch.object(charm.hookenv, "open_port")
+    mock_close_port = mocker.patch.object(charm.hookenv, "close_port")
+
+    with harness.hooks_disabled():
+        harness.update_config({"scrape-port": new_port})
+
+    harness.charm.reconfigure_open_ports()
+
+    mock_close_port.assert_called_once_with(old_port, old_protocol)
+    mock_open_port.assert_called_once_with(new_port)
+
+
+@pytest.mark.parametrize("error", [True, False])
+def test_on_install_callback(error, harness, mocker):
+    """Test handling of InstallEvent with '_on_install' callback."""
+    snap_exception = charm.snap.CouldNotAcquireLockException
+    exporter_install = mocker.patch.object(harness.charm.exporter, "install")
+
+    if error:
```

Comment:
I think that you forgot to split also this function.

Reply:
Do you mean we should split the success and fail case ?  I am curious that what is the benefit for unit test?

Reply:
I agree in this case with Robert. I've split other tests before but forgot about this one. The idea is that each unit test should focus on one specific thing and having `if/else` branch in the test that verifies different things can make it hard to read.
On top of that, having two tests with verbose names
- `test_on_install_callback_success`
- `test_on_install_callback_fail`

can make the interpretation of results easier.